core/src/main/scala/cromwell/util/JsonEditor.scala

```diff
@@ -72,21 +73,77 @@ object JsonEditor {
 
   def logs(json: Json): Json = includeJson(json, NonEmptyList.of("stdout", "stderr", "backendLogs", "id"))
 
+  implicit class EnhancedJson(val json: Json) extends AnyVal {
+    def rootWorkflowId: Option[WorkflowId] = {
+      for {
+        o <- json.asObject
+        id <- o.kleisli("id")
+        s <- id.asString
+      } yield WorkflowId.fromString(s)
+    }
+  }
+
   /**
-    * We only return labels associated with top-level workflows.  Subworkflows don't include labels (as of 7/26/19).
-    *
-    * Thus this method puts the labels as a top-level field.
+    * In-memory upsert of labels into the base Json, handling root and sub workflows appropriately.
     *
     * @param json json blob with or without "labels" field
-    * @param labels a map of labels one would like to apply to a workflow json
+    * @param databaseLabels a map of workflow IDs to maps of labels one would like to apply to a workflow json
     * @return json with labels merged in.  Any prior non-object "labels" field will be overwritten and any object fields will be merged together and - again - any existing values overwritten.
     */
-  def augmentLabels(json: Json, labels: Map[String, String]): Json = {
-    val newData: Json = Json.fromFields(labels.safeMapValues(Json.fromString))
-    val newObj: Json = Json.fromFields(List(("labels", newData)))
-    //in the event of a key clash, the values in "newObj" will be favored over "json"
-    json deepMerge newObj
-  }
+  def updateLabels(json: Json, databaseLabels: Map[WorkflowId, Map[String, String]]): Json = {
+    val subWorkflowMetadataKey = "subWorkflowMetadata"
+
+    def doUpdateWorkflow(workflowJson: Json): Json = {
+      val id: String = workflowJson.rootWorkflowId.
+        getOrElse(throw new RuntimeException(s"did not find workflow id in ${workflowJson.printWith(Printer.spaces2)}")).toString
```

Comment:
Might be nice to include the target WF ID
```suggestion
        getOrElse(throw new RuntimeException(s"did not find workflow $id in ${workflowJson.printWith(Printer.spaces2)}")).toString
```

Also, isn't it potentially dangerous to print out an entire metadata JSON to the log? Can we take the first 100 chars or something?

Reply:
This code actually doesn't know the target workflow id, it's trying to read it from the JSON. Agree with the point about dumping the whole workflow, will fix.